File: Searching/Sahil/C++/jump_search.cpp
/*
Works only with sorted arrays
*/

#include<iostream>
#include<array>
#include<bits/stdc++.h>
using namespace std;

int jumpSearch(int arr[], int size, int number_to_search) {

    int step = sqrt(size) ;
    int totalStep = (size) / sqrt(size)  ;
    int previous_step = 0 ;

    while (arr[step] < number_to_search && step < size) { 

        // searching for the suitable block
        previous_step = step;
        step += sqrt(size) ;
    } 
    while (arr[previous_step] < number_to_search) {

        // linear search in that block till that element is reached
        previous_step += 1 ;
        if(previous_step == size) {
            return -1 ;
        }
    }
    // checking if that element is the element we require
    if(arr[previous_step] == number_to_search) {

        return previous_step ;
    }

    return -1 ;
}

int main() {

    array<int, 16> array = { 1, 2 , 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16} ;
    int number_to_search = 14 ;

    int search_result = jumpSearch(array.data(), array.size(), number_to_search) ;

    if (search_result >= 0) {

        cout << "got it at: " << search_result + 1 << endl ;
    } else {

        cout << "No such value present" << endl ;
    }
    return 0 ;
}

File: Searching/Sahil/JAVA/JumpSearch.java
package searching;

public class JumpSearch {

    public static int jumpSearch(int[] arr, int numberToSearch) {

        int size = arr.length;
        int step = (int) Math.sqrt(size) ;
        int previous_step = 0 ;

        while (step < size && arr[step] < numberToSearch) {

            // searching for the suitable block
            previous_step = step;
            step += Math.sqrt(size) ;
        }
        while (arr[previous_step] < numberToSearch) {

            // linear search in that block till that element is reached
            previous_step += 1 ;
            if(previous_step == size) {
                return -1 ;
            }
        }
        // checking if that element is the element we require
        if(arr[previous_step] == numberToSearch) {

            return previous_step ;
        }

        return -1 ;
    }

    public static void main(String[] args) {

        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200 } ;
        int numberToSearch = 190 ;
        int searchResult = jumpSearch(arr, numberToSearch) ;

        if(searchResult >= 0) {
            System.out.println("Number found at: " + (searchResult + 1));
        } else {
            System.out.println("Number not found");
        }
    }
}
